### PR TITLE
Add AnyView with HasFromAnyView instances

### PR DESCRIPTION
**Pull request by dylant-da**
https://github.com/digital-asset/ghc/pull/138

part of #14830

### CODE REVIEW

File: compiler/damlc/daml-stdlib-src/DA/Internal/Interface.daml [hunk 1]
```diff
@@ -175,3 +179,30 @@ class HasInterfaceView i v | i -> v where
...
 view : forall i v. HasInterfaceView i v => i -> v
 view = _view
+
+class HasFromAnyView i v | i -> v where
+  -- | HIDE
+  _fromAnyView : proxy i -> Any -> Optional v
+
+class HasToAnyView i v | i -> v where
+  -- | HIDE
+  _toAnyView : proxy i -> v -> Any
```
Comment (akrmn) [inline compiler/damlc/daml-stdlib-src/DA/Internal/Interface.daml:189]: As discussed last week, we don't expect client code to construct views, so this class is unnecessary

File: compiler/damlc/daml-stdlib-src/DA/Internal/Interface/AnyView.daml [hunk 2]
```diff
@@ -0,0 +1,19 @@
+-- Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+
+module DA.Internal.Interface.AnyView (InterfaceTypeRep (..), AnyView (..)) where
```
Comment (akrmn) [inline compiler/damlc/daml-stdlib-src/DA/Internal/Interface/AnyView.daml:7]: I like the approach of other `StablePackages` where the stable module is named `Foo.Bar.Types` and the non-stable just `Foo.Bar`

File: compiler/damlc/daml-stdlib-src/DA/Internal/Interface.daml [hunk 3]
```diff
@@ -175,3 +179,30 @@ class HasInterfaceView i v | i -> v where
...
+  | otherwise = None
+
+toAnyView : forall i v. (HasTemplateTypeRep i, HasToAnyView i v) => v -> AnyView
+toAnyView v =
+  AnyView
+    (_toAnyView ([] : [i]) v)
+    (fromTemplateTypeRep (templateTypeRep @i))
+
+deriving instance Eq InterfaceTypeRep
+deriving instance Ord InterfaceTypeRep
```
Comment (akrmn) [inline compiler/damlc/daml-stdlib-src/DA/Internal/Interface.daml:204]: Following from [this comment](https://github.com/digital-asset/daml/pull/15280/files#r1003072116), these instances would go under `DA.Internal.Interface.AnyView`

Comment (akrmn) [reply compiler/damlc/daml-stdlib-src/DA/Internal/Interface.daml:204]: We should also add the appropriate `HasField` instances under `DA.Internal.Interface.AnyView`

**Review by akrmn — APPROVED** [review]
LGTM other than:
* the unnecessary `HasToAnyView` class and instances
* I'd rather define `newtype InterfaceTypeRep` and `data AnyView` under a stable package for module `DA.Internal.Interface.AnyView.Types`, with the `Eq`, `Ord` and `HasField` instances under non-stable module `DA.Internal.Interface.AnyView`. The other additions to `DA.Internal.Interface` (excluding `HasToAnyView`/`toAnyView`) would stay there.

**Comment by dylant-da** [comment]
Sounds good, will do :)

File: compiler/damlc/daml-stdlib-src/DA/Internal/Interface.daml (hunk 1 above)
Comment (dylant-da) [reply compiler/damlc/daml-stdlib-src/DA/Internal/Interface.daml]: done

File: compiler/damlc/stable-packages/lib/DA/Daml/StablePackages.hs [hunk 4]
```diff
@@ -188,6 +189,29 @@ daInternalAny = package version1_7 $ NM.singleton (emptyModule modName)
           DataRecord [(mkField "getAnyContractKey", TAny), (mkField "getAnyContractKeyTemplateTypeRep", TCon (Qualified PRSelf modName (mkTypeCon ["TemplateTypeRep"])))]
       ]
 
+daInternalInterfaceAnyView :: Package
+daInternalInterfaceAnyView = Package
+  { packageLfVersion = version1_15
+  , packageModules = NM.singleton (emptyModule modName)
+      { moduleDataTypes = datatypes
+      }
```
Comment (akrmn) [inline compiler/damlc/stable-packages/lib/DA/Daml/StablePackages.hs:197]: I forgot to mention earlier, we should also add the appropriate selector and worker definitions under `moduleValues`,

```suggestion
daInternalInterfaceAnyView = Package
  { packageLfVersion = version1_15
  , packageModules = NM.singleton (emptyModule modName)
      { moduleDataTypes = datatypes
      { moduleValues = values
      }
```

File: compiler/damlc/daml-stdlib-src/DA/Internal/Interface.daml [hunk 5]
```diff
@@ -4,6 +4,7 @@
 {-# LANGUAGE NoImplicitPrelude #-}
 {-# LANGUAGE ScopedTypeVariables #-}
 {-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE StandaloneDeriving #-}
```
Comment (akrmn) [inline compiler/damlc/daml-stdlib-src/DA/Internal/Interface.daml:7]: No longer needed

```suggestion
```

File: compiler/damlc/daml-stdlib-src/DA/Internal/Interface/AnyView.daml (hunk 2 above)
Comment (dylant-da) [reply compiler/damlc/daml-stdlib-src/DA/Internal/Interface/AnyView.daml]: done

File: compiler/damlc/daml-stdlib-src/DA/Internal/Interface.daml (hunk 3 above)
Comment (dylant-da) [reply compiler/damlc/daml-stdlib-src/DA/Internal/Interface.daml]: done

File: compiler/damlc/stable-packages/lib/DA/Daml/StablePackages.hs (hunk 4 above)
Comment (dylant-da) [reply compiler/damlc/stable-packages/lib/DA/Daml/StablePackages.hs]: done

**Comment by remyhaemmerle-da** [comment]
cc @carlpulley-da. 
If you need any help using that, please ping me.

**Pull request merged**